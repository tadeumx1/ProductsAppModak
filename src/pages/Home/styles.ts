import { StyleSheet } from 'react-native';

export const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  containerProducts: {
    flex: 1,
    backgroundColor: '#f5f5f5',
    padding: 16,
  },
  list: {
    paddingBottom: 16,
  },
  loadingErrorContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    padding: 16,
    backgroundColor: '#f5f5f5',
  },
  titleLoading: {
    marginTop: 16,
    fontSize: 18,
    fontWeight: 'bold',
    color: '#6200ee',
  },
  errorText: {
    fontSize: 16,
    color: '#b00020',
    marginBottom: 16,
  },
  card: {
    marginBottom: 16,
    borderRadius: 8,
    elevation: 2,
  },
  loadingMoreContainer: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    paddingVertical: 16,
  },
  loadingMoreText: {
    marginLeft: 8,
    fontSize: 14,
    color: '#6200ee',
  },
  header: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    paddingHorizontal: 16,
    paddingVertical: 12,
    backgroundColor: '#f5f5f5',
    borderBottomWidth: 1,
    borderBottomColor: '#e0e0e0',
  },
  headerTitle: {
    fontSize: 24,
    fontWeight: 'bold',
    fontFamily: 'Roboto',
    color: '#333333',
  },
  menuItem: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  filterButtonContent: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  filterButton: {
    justifyContent: 'center',
  },
  emptyListContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    padding: 16,
  },
  emptyListText: {
    fontSize: 16,
    color: '#888',
    textAlign: 'center',
  },
  sortButtonContent: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  sortButton: {
    justifyContent: 'center',
  },
});
