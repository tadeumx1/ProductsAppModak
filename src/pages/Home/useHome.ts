import { useEffect, useMemo, useRef, useState } from 'react';
import {
  useGetProductCategories,
  useGetProductsData,
} from '../../services/modules/product';
import { RouteProp, useNavigation, useRoute } from '@react-navigation/native';
import { NativeStackNavigationProp } from '@react-navigation/native-stack';
import notifee from '@notifee/react-native';
import { ProductObject } from '../../services/modules/product/types';
import { RootParamListStack } from '@/src/types';

type HomeNavigationProp = NativeStackNavigationProp<RootParamListStack, 'Home'>;

type HomeRouteProp = RouteProp<
  { Home: { categoryName: string | null } },
  'Home'
>;

export const useHome = () => {
  const limit = 5;

  const [categorySelected, setCategorySelected] = useState<string | null>(null);
  const [filterModalVisible, setFilterModalVisible] = useState(false);
  const [filterMenuVisible, setFilterMenuVisible] = useState(false);
  const [sortMenuVisible, setSortMenuVisible] = useState(false);
  const [sortCriteria, setSortCriteria] = useState<string | null>(null);
  const route = useRoute<HomeRouteProp>();
  const { categoryName } = route.params || {};
  const onEndReachedCalledDuringMomentum = useRef(false);

  const {
    data,
    isLoading: isLoadingGetProductsData,
    isError: isErrorGetProductsData,
    hasNextPage,
    fetchNextPage,
    refetch: getProductsData,
    isFetchingNextPage,
  } = useGetProductsData({
    limit,
    category: categorySelected,
  });

  const {
    data: dataProductCategories,
    isLoading: isLoadingProductCategories,
    isError: isErrorProductCategories,
    refetch: getProductCategories,
  } = useGetProductCategories();

  const navigation = useNavigation<HomeNavigationProp>();

  const openFilterModal = () => setFilterModalVisible(true);
  const closeFilterModal = () => setFilterModalVisible(false);
  const toggleFilterMenu = () => setFilterMenuVisible((prev) => !prev);
  const closeFilterMenu = () => setFilterMenuVisible(false);
  const openSortMenu = () => setSortMenuVisible(true);
  const closeSortMenu = () => setSortMenuVisible(false);

  const handleGetNotificationPermission = async () => {
    await notifee.requestPermission();
  };

  useEffect(() => {
    handleGetNotificationPermission();
  }, []);

  const pages = (data?.pages?.length || 0) > 0;

  useEffect(() => {
    const isNavigatedFromDeepLink =
      categoryName !== null && route.name === 'Home';

    if (
      isNavigatedFromDeepLink &&
      pages &&
      !isLoadingGetProductsData &&
      !isErrorGetProductsData
    ) {
      setCategorySelected(categoryName);
    }
  }, [pages, categoryName, isLoadingGetProductsData, isErrorGetProductsData]);

  const filterSortProducts = useMemo(() => {
    if (!data) return [];

    let allProducts = data.pages.flatMap((page) => page.products);

    if (!categorySelected) getProductsData();

    if (categorySelected) {
      allProducts = allProducts.filter(
        (product) => product.category === categorySelected
      );
    }

    if (sortCriteria === 'price') {
      allProducts = allProducts.sort((a, b) => a.price - b.price);
    } else if (sortCriteria === 'rating') {
      allProducts = allProducts.sort((a, b) => b.rating - a.rating);
    }

    return allProducts;
  }, [data, categorySelected, sortCriteria]);

  const handleSortCriteria = (criteria: string) => {
    if (sortCriteria === criteria) {
      setSortCriteria(null);
    } else {
      setSortCriteria(criteria);
    }

    closeSortMenu();
  };

  const handleNavigateToProductDetail = (product: ProductObject) => {
    navigation.navigate('ProductDetail', { product });
  };

  return {
    products: filterSortProducts,
    categoriesProducts: dataProductCategories,
    isLoadingGetProductsData,
    isErrorGetProductsData,
    isLoadingProductCategories,
    isErrorProductCategories,
    fetchNextPage,
    isFetchingNextPage,
    hasNextPage,
    toggleFilterMenu,
    closeFilterMenu,
    openSortMenu,
    closeSortMenu,
    setSortCriteria,
    categorySelected,
    setCategorySelected,
    filterMenuVisible,
    sortMenuVisible,
    filterModalVisible,
    limit,
    setFilterModalVisible,
    openFilterModal,
    closeFilterModal,
    handleNavigateToProductDetail,
    sortCriteria,
    onEndReachedCalledDuringMomentum,
    handleSortCriteria,
    getProductsData,
    getProductCategories,
  };
};
