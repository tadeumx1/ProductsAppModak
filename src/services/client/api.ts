import axios from 'axios';

const url = 'https://dummyjson.com';

export const api = axios.create({
  baseURL: url,
});
