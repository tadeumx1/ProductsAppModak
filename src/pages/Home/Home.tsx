import React from 'react';
import {
  FlatList,
  Text,
  View,
  ActivityIndicator,
  SafeAreaView,
  TouchableOpacity,
} from 'react-native';
import { Button, Menu } from 'react-native-paper';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';
import { useHome } from './useHome';
import { Product } from '../../components/Product';
import { CategoriesModal } from '@/src/components/CategoriesModal';
import { styles } from './styles';
import { ErrorContainer } from '@/src/components/ErrorContainer';

const Home = () => {
  const {
    products,
    isErrorGetProductsData,
    isLoadingGetProductsData,
    isLoadingProductCategories,
    isErrorProductCategories,
    fetchNextPage,
    isFetchingNextPage,
    hasNextPage,
    sortMenuVisible,
    openSortMenu,
    closeSortMenu,
    categorySelected,
    setCategorySelected,
    categoriesProducts,
    handleNavigateToProductDetail,
    filterModalVisible,
    openFilterModal,
    closeFilterModal,
    sortCriteria,
    onEndReachedCalledDuringMomentum,
    handleSortCriteria,
    getProductsData,
    getProductCategories,
  } = useHome();

  if (isLoadingGetProductsData) {
    return (
      <View testID="loading-container" style={styles.loadingErrorContainer}>
        <ActivityIndicator size="large" color="#6200ee" />
        <Text>Loading Products</Text>
      </View>
    );
  }

  if (isLoadingProductCategories) {
    return (
      <View testID="loading-container" style={styles.loadingErrorContainer}>
        <ActivityIndicator size="large" color="#6200ee" />
        <Text>Loading</Text>
      </View>
    );
  }

  if (isErrorGetProductsData) {
    return (
      <ErrorContainer
        errorOperation="Products"
        onRetry={() => getProductsData()}
      />
    );
  }

  if (isErrorProductCategories) {
    return (
      <ErrorContainer
        errorOperation="Product Categories"
        onRetry={() => getProductCategories()}
      />
    );
  }

  return (
    <SafeAreaView style={styles.container}>
      <View style={styles.header}>
        <Text style={styles.headerTitle}>ProductsApp</Text>
        <Button
          mode="outlined"
          icon="filter"
          onPress={openFilterModal}
          style={styles.filterButton}
          contentStyle={styles.filterButtonContent}
        >
          <Text>Filter</Text>
        </Button>
        <Menu
          visible={sortMenuVisible}
          onDismiss={closeSortMenu}
          anchor={
            <Button
              mode="outlined"
              icon="sort"
              onPress={openSortMenu}
              style={styles.sortButton}
              contentStyle={styles.sortButtonContent}
            >
              <Text>Sort</Text>
            </Button>
          }
        >
          <Menu.Item
            onPress={() => handleSortCriteria('price')}
            title={
              <View style={styles.menuItem}>
                <Text>Sort by Price</Text>
                {sortCriteria === 'price' && (
                  <Icon name="check" size={20} color="#6200ee" />
                )}
              </View>
            }
          />
          <Menu.Item
            onPress={() => handleSortCriteria('rating')}
            title={
              <View style={styles.menuItem}>
                <Text>Sort by Rating</Text>
                {sortCriteria === 'rating' && (
                  <Icon name="check" size={20} color="#6200ee" />
                )}
              </View>
            }
          />
        </Menu>
      </View>
      <View style={styles.containerProducts}>
        <FlatList
          data={products}
          keyExtractor={(item) => item.id.toString()}
          renderItem={({ item }) => (
            <TouchableOpacity
              onPress={() => handleNavigateToProductDetail(item)}
            >
              <Product
                title={item.title}
                thumbnail={item.thumbnail}
                price={item.price}
              />
            </TouchableOpacity>
          )}
          contentContainerStyle={{
            flexGrow: 1,
          }}
          onEndReached={() => {
            if (
              hasNextPage &&
              !isFetchingNextPage &&
              !onEndReachedCalledDuringMomentum.current
            ) {
              onEndReachedCalledDuringMomentum.current = true;
              fetchNextPage();
            }
          }}
          onMomentumScrollBegin={() => {
            onEndReachedCalledDuringMomentum.current = false;
          }}
          onEndReachedThreshold={0.1}
          testID="product-list"
          ListEmptyComponent={() => (
            <View style={styles.emptyListContainer}>
              <Text style={styles.emptyListText}>No products available</Text>
            </View>
          )}
          ListFooterComponent={() =>
            isFetchingNextPage ? (
              <View style={styles.loadingMoreContainer}>
                <ActivityIndicator size="small" color="#6200ee" />
                <Text style={styles.loadingMoreText}>Loading more...</Text>
              </View>
            ) : null
          }
        />
        <CategoriesModal
          filterModalVisible={filterModalVisible}
          closeFilterModal={closeFilterModal}
          categoriesProducts={categoriesProducts}
          categorySelected={categorySelected}
          setCategorySelected={setCategorySelected}
        />
      </View>
    </SafeAreaView>
  );
};

export default Home;
