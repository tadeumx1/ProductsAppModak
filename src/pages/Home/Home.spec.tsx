import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react-native';
import { Provider as PaperProvider, DefaultTheme } from 'react-native-paper';
import Home from './Home';
import { useHome } from './useHome';
import { api } from '../../services/client';

jest.mock('./useHome');
const mockedUseHome = useHome as jest.Mock;

jest.mock('@notifee/react-native', () => ({
  requestPermission: jest.fn(),
  displayNotification: jest.fn(),
  createChannel: jest.fn(),
}));

jest.mock('../../services/client/api');
const mockedApi = api as jest.Mocked<typeof api>;

mockedApi.get.mockResolvedValue({});

const mockResponsePage1 = {
  products: [
    { id: 1, title: 'Product 1', thumbnail: '', price: 10 },
    { id: 2, title: 'Product 2', thumbnail: '', price: 20 },
  ],
  total: 5,
};

jest.mock('react-native-paper', () => {
  const React = require('react');
  const { View, Text, TouchableOpacity } = require('react-native');
  const actualPaper = jest.requireActual('react-native-paper');

  return {
    ...actualPaper,
    PaperProvider: ({ children }: any) => <View>{children}</View>,
    ActivityIndicator: () => <View />,
    Text: ({ children }: { children: string }) => <Text>{children}</Text>,
    Button: jest.fn().mockImplementation(({ children, ...props }) => (
      <TouchableOpacity {...props}>
        <Text>{children}</Text>
      </TouchableOpacity>
    )),
    Menu: ({ visible, onDismiss, anchor, children }: any) => (
      <View>
        {anchor}
        {visible && <View>{children}</View>}
      </View>
    ),
    MenuItem: ({ title, onPress }: { title: string; onPress: () => void }) => (
      <Text onPress={onPress}>{title}</Text>
    ),
    Portal: ({ children }: any) => <View>{children}</View>,
    Modal: ({ visible, children }: any) =>
      visible ? <View>{children}</View> : null,
  };
});

const renderWithProvider = (ui: React.ReactElement) => {
  return render(<PaperProvider theme={DefaultTheme}>{ui}</PaperProvider>);
};

describe('Home Component', () => {
  const mockNavigate = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    mockedUseHome.mockReturnValue({
      products: [],
      isError: false,
      isLoading: false,
      fetchNextPage: jest.fn(),
      isFetchingNextPage: false,
      hasNextPage: true,
      sortMenuVisible: false,
      openSortMenu: jest.fn(),
      closeSortMenu: jest.fn(),
      categorySelected: null,
      setCategorySelected: jest.fn(),
      categoriesProducts: [],
      handleNavigateToProductDetail: mockNavigate,
      filterModalVisible: false,
      openFilterModal: jest.fn(),
      closeFilterModal: jest.fn(),
      sortCriteria: null,
      onEndReachedCalledDuringMomentum: { current: false },
      handleSortCriteria: jest.fn(),
    });
  });

  it('renders loading state', async () => {
    mockedUseHome.mockReturnValue({
      ...mockedUseHome(),
      isLoadingGetProductsData: true,
      isErrorGetProductsData: false,
      products: [],
    });

    const { getByText, toJSON, getByTestId } = renderWithProvider(<Home />);

    expect(getByTestId('loading-container')).toBeDefined();
    expect(getByText('Loading Products')).toBeDefined();
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders error state', async () => {
    mockedUseHome.mockReturnValue({
      ...mockedUseHome(),
      isLoadingGetProductsData: false,
      isErrorGetProductsData: true,
      products: mockResponsePage1,
    });

    const { getByText, toJSON } = renderWithProvider(<Home />);

    await waitFor(() => {
      expect(getByText('Something Went Wrong')).toBeDefined();
      expect(
        getByText('Please check your connection and try again')
      ).toBeDefined();
      expect(getByText('Products')).toBeDefined();
      expect(getByText('Retry')).toBeDefined();
      expect(toJSON()).toMatchSnapshot();
    });
  });

  it('renders empty list message when no products are available', async () => {
    const { getByText, toJSON } = renderWithProvider(<Home />);

    await waitFor(() => {
      expect(getByText('No products available')).toBeDefined();
      expect(toJSON()).toMatchSnapshot();
    });
  });

  it('renders products correctly', async () => {
    mockedUseHome.mockReturnValue({
      ...mockedUseHome(),
      products: [
        {
          id: 1,
          title: 'Product 1',
          thumbnail: 'https://example.com/1.jpg',
          price: 19.99,
        },
        {
          id: 2,
          title: 'Product 2',
          thumbnail: 'https://example.com/2.jpg',
          price: 29.99,
        },
      ],
    });

    const { getByText, toJSON } = renderWithProvider(<Home />);

    await waitFor(() => {
      expect(getByText('Product 1')).toBeDefined();
      expect(getByText('$19.99')).toBeDefined();
      expect(getByText('Product 2')).toBeDefined();
      expect(getByText('$29.99')).toBeDefined();

      expect(toJSON()).toMatchSnapshot();
    });
  });

  it('navigates to product detail on product press', () => {
    mockedUseHome.mockReturnValue({
      ...mockedUseHome(),
      products: [
        {
          id: 1,
          title: 'Product 1',
          thumbnail: 'https://example.com/1.jpg',
          price: 19.99,
        },
      ],
    });

    const { getByText } = renderWithProvider(<Home />);
    const product = getByText('Product 1');

    fireEvent.press(product);

    expect(mockNavigate).toHaveBeenCalledWith({
      id: 1,
      title: 'Product 1',
      thumbnail: 'https://example.com/1.jpg',
      price: 19.99,
    });
  });
});
