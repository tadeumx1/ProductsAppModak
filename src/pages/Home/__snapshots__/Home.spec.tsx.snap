// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component renders empty list message when no products are available 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    {
      "flex": 1,
    }
  }
>
  <RCTSafeAreaView
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#f5f5f5",
          "borderBottomColor": "#e0e0e0",
          "borderBottomWidth": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingHorizontal": 16,
          "paddingVertical": 12,
        }
      }
    >
      <Text
        style={
          {
            "color": "#333333",
            "fontFamily": "Roboto",
            "fontSize": 24,
            "fontWeight": "bold",
          }
        }
      >
        ProductsApp
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "justifyContent": "center",
            "opacity": 1,
          }
        }
      >
        <Text>
          <Text>
            Filter
          </Text>
        </Text>
      </View>
      <View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "justifyContent": "center",
              "opacity": 1,
            }
          }
        >
          <Text>
            <Text>
              Sort
            </Text>
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "backgroundColor": "#f5f5f5",
          "flex": 1,
          "padding": 16,
        }
      }
    >
      <RCTScrollView
        ListEmptyComponent={[Function]}
        ListFooterComponent={[Function]}
        contentContainerStyle={
          {
            "flexGrow": 1,
          }
        }
        data={[]}
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onEndReached={[Function]}
        onEndReachedThreshold={0.1}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={0.0001}
        stickyHeaderIndices={[]}
        testID="product-list"
        viewabilityConfigCallbackPairs={[]}
      >
        <View>
          <View
            style={
              {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "center",
                "padding": 16,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#888",
                  "fontSize": 16,
                  "textAlign": "center",
                }
              }
            >
              No products available
            </Text>
          </View>
          <View
            onLayout={[Function]}
          />
        </View>
      </RCTScrollView>
    </View>
  </RCTSafeAreaView>
</View>
`;

exports[`Home Component renders error state 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#F9F9F9",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        [
          {
            "color": "#D32F2F",
            "fontSize": 80,
          },
          undefined,
          {
            "fontFamily": "Material Design Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
    >
      ó°—–
    </Text>
    <Text
      style={
        {
          "color": "#333",
          "fontSize": 24,
          "fontWeight": "bold",
          "marginBottom": 8,
          "marginTop": 16,
        }
      }
    >
      Something Went Wrong
    </Text>
    <Text
      style={
        {
          "color": "##7d7c7c",
          "fontSize": 16,
          "marginBottom": 24,
          "paddingHorizontal": 16,
          "textAlign": "center",
        }
      }
    >
      Please check your connection and try again
    </Text>
    <Text
      style={
        {
          "color": "##7d7c7c",
          "fontSize": 16,
          "marginBottom": 24,
          "paddingHorizontal": 16,
          "textAlign": "center",
        }
      }
    >
      Products
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#6200EE",
          "borderRadius": 4,
          "height": undefined,
          "justifyContent": "center",
          "marginHorizontal": 8,
          "opacity": 1,
          "width": undefined,
        }
      }
    >
      <Text>
        <Text>
          Retry
        </Text>
      </Text>
    </View>
  </View>
</View>
`;

exports[`Home Component renders loading state 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#f5f5f5",
        "flex": 1,
        "justifyContent": "center",
        "padding": 16,
      }
    }
    testID="loading-container"
  >
    <ActivityIndicator
      color="#6200ee"
      size="large"
    />
    <Text>
      Loading Products
    </Text>
  </View>
</View>
`;

exports[`Home Component renders products correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    {
      "flex": 1,
    }
  }
>
  <RCTSafeAreaView
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#f5f5f5",
          "borderBottomColor": "#e0e0e0",
          "borderBottomWidth": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingHorizontal": 16,
          "paddingVertical": 12,
        }
      }
    >
      <Text
        style={
          {
            "color": "#333333",
            "fontFamily": "Roboto",
            "fontSize": 24,
            "fontWeight": "bold",
          }
        }
      >
        ProductsApp
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "justifyContent": "center",
            "opacity": 1,
          }
        }
      >
        <Text>
          <Text>
            Filter
          </Text>
        </Text>
      </View>
      <View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "justifyContent": "center",
              "opacity": 1,
            }
          }
        >
          <Text>
            <Text>
              Sort
            </Text>
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "backgroundColor": "#f5f5f5",
          "flex": 1,
          "padding": 16,
        }
      }
    >
      <RCTScrollView
        ListEmptyComponent={[Function]}
        ListFooterComponent={[Function]}
        contentContainerStyle={
          {
            "flexGrow": 1,
          }
        }
        data={
          [
            {
              "id": 1,
              "price": 19.99,
              "thumbnail": "https://example.com/1.jpg",
              "title": "Product 1",
            },
            {
              "id": 2,
              "price": 29.99,
              "thumbnail": "https://example.com/2.jpg",
              "title": "Product 2",
            },
          ]
        }
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onEndReached={[Function]}
        onEndReachedThreshold={0.1}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={0.0001}
        stickyHeaderIndices={[]}
        testID="product-list"
        viewabilityConfigCallbackPairs={[]}
      >
        <View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={null}
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  {
                    "backgroundColor": "rgba(247, 243, 249, 1)",
                    "borderRadius": 8,
                    "marginBottom": 16,
                    "shadowColor": "#000",
                    "shadowOffset": {
                      "height": 1,
                      "width": 0,
                    },
                    "shadowOpacity": 0.15,
                    "shadowRadius": 3,
                  }
                }
                testID="card-container-outer-layer"
              >
                <View
                  collapsable={false}
                  style={
                    {
                      "backgroundColor": "rgba(247, 243, 249, 1)",
                      "borderRadius": 8,
                      "elevation": 3,
                      "flex": undefined,
                      "shadowColor": "#000",
                      "shadowOffset": {
                        "height": 1,
                        "width": 0,
                      },
                      "shadowOpacity": 0.3,
                      "shadowRadius": 1,
                    }
                  }
                  testID="card-container"
                >
                  <View
                    style={
                      [
                        {
                          "flexShrink": 1,
                        },
                        undefined,
                      ]
                    }
                    testID="card"
                  >
                    <View
                      borderRadiusStyles={
                        {
                          "borderRadius": 8,
                        }
                      }
                      style={
                        [
                          {
                            "paddingHorizontal": 16,
                          },
                          {
                            "paddingVertical": 16,
                          },
                          {
                            "alignItems": "center",
                            "flexDirection": "row",
                          },
                        ]
                      }
                    >
                      <View
                        style={
                          [
                            {
                              "overflow": "hidden",
                            },
                            {
                              "borderRadius": 25,
                              "height": 64,
                              "marginRight": 10,
                              "width": 64,
                            },
                          ]
                        }
                      >
                        <FastImageView
                          defaultSource={null}
                          resizeMode="contain"
                          source={
                            {
                              "priority": "normal",
                              "uri": "https://example.com/1.jpg",
                            }
                          }
                          style={
                            {
                              "bottom": 0,
                              "left": 0,
                              "position": "absolute",
                              "right": 0,
                              "top": 0,
                            }
                          }
                          testID="image-thumbnail-container"
                        />
                      </View>
                      <View
                        style={
                          {
                            "flex": 1,
                            "justifyContent": "center",
                          }
                        }
                      >
                        <Text>
                          Product 1
                        </Text>
                        <Text>
                          $
                          19.99
                        </Text>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={null}
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  {
                    "backgroundColor": "rgba(247, 243, 249, 1)",
                    "borderRadius": 8,
                    "marginBottom": 16,
                    "shadowColor": "#000",
                    "shadowOffset": {
                      "height": 1,
                      "width": 0,
                    },
                    "shadowOpacity": 0.15,
                    "shadowRadius": 3,
                  }
                }
                testID="card-container-outer-layer"
              >
                <View
                  collapsable={false}
                  style={
                    {
                      "backgroundColor": "rgba(247, 243, 249, 1)",
                      "borderRadius": 8,
                      "elevation": 3,
                      "flex": undefined,
                      "shadowColor": "#000",
                      "shadowOffset": {
                        "height": 1,
                        "width": 0,
                      },
                      "shadowOpacity": 0.3,
                      "shadowRadius": 1,
                    }
                  }
                  testID="card-container"
                >
                  <View
                    style={
                      [
                        {
                          "flexShrink": 1,
                        },
                        undefined,
                      ]
                    }
                    testID="card"
                  >
                    <View
                      borderRadiusStyles={
                        {
                          "borderRadius": 8,
                        }
                      }
                      style={
                        [
                          {
                            "paddingHorizontal": 16,
                          },
                          {
                            "paddingVertical": 16,
                          },
                          {
                            "alignItems": "center",
                            "flexDirection": "row",
                          },
                        ]
                      }
                    >
                      <View
                        style={
                          [
                            {
                              "overflow": "hidden",
                            },
                            {
                              "borderRadius": 25,
                              "height": 64,
                              "marginRight": 10,
                              "width": 64,
                            },
                          ]
                        }
                      >
                        <FastImageView
                          defaultSource={null}
                          resizeMode="contain"
                          source={
                            {
                              "priority": "normal",
                              "uri": "https://example.com/2.jpg",
                            }
                          }
                          style={
                            {
                              "bottom": 0,
                              "left": 0,
                              "position": "absolute",
                              "right": 0,
                              "top": 0,
                            }
                          }
                          testID="image-thumbnail-container"
                        />
                      </View>
                      <View
                        style={
                          {
                            "flex": 1,
                            "justifyContent": "center",
                          }
                        }
                      >
                        <Text>
                          Product 2
                        </Text>
                        <Text>
                          $
                          29.99
                        </Text>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
          <View
            onLayout={[Function]}
          />
        </View>
      </RCTScrollView>
    </View>
  </RCTSafeAreaView>
</View>
`;
